src/types/capsule.rs

```diff
@@ -89,36 +89,39 @@ impl PyCapsule {
     ///
     /// Also provides a destructor: when the `PyCapsule` is destroyed, it will be passed the original object,
     /// as well as a `*mut c_void` which will point to the capsule's context, if any.
+    ///
+    /// The `destructor` must be `Send`, because there is no guarantee which thread it will eventually
+    /// be called from.
     pub fn new_with_destructor<
         'py,
         T: 'static + Send + AssertNotZeroSized,
-        F: FnOnce(T, *mut c_void),
+        F: FnOnce(T, *mut c_void) + Send,
     >(
         py: Python<'py>,
         value: T,
-        name: &CStr,
+        name: Option<CString>,
         destructor: F,
     ) -> PyResult<&'py Self> {
         AssertNotZeroSized::assert_not_zero_sized(&value);
-        // Take ownership of a copy of `name` so that the string is guaranteed to live as long
-        // as the capsule. PyCapsule_new purely saves the pointer in the capsule so doesn't
-        // guarantee ownership itself.
-        let name = name.to_owned();
-        let name_ptr = name.as_ptr();
+
+        // Sanity check for capsule layout
+        debug_assert_eq!(memoffset::offset_of!(CapsuleContents::<T, F>, value), 0);
+
+        let name_ptr = name.as_ref().map_or(std::ptr::null(), |name| name.as_ptr());
```

Comment:
I think this could also use `name_as_ptr`, probably via `name_as_ptr(name.as_deref())` (which should be in MSRV reach at 1.40)?